lib/ecto/repo/schema.ex

```diff
@@ -116,6 +116,19 @@ defmodule Ecto.Repo.Schema do
     end
 
     header = case query.select do
+      %Ecto.Query.SelectExpr{expr: {:%{}, [], [{:|, _, [{:&, _, [ix]}, args]}]}, fields: fields} ->
+        select_fields = for {{:., _, [{:&, _, [^ix]}, field]}, [], []} <- fields, do: field
+        update_fields = Keyword.keys(args)
+
+        select_fields ++ update_fields
```

Comment:
Do we need to solve duplicates somehow?

We need a test for `%{map(p, [:lock_version]) | lock_version: 2}` and make sure the update wins?

Reply:
Btw, whenever possible, I prefer unit tests, because it has a way quicker feedback loop. I am not sure if we are unit testing this functionality though (we might!).